Docs: Document `global-not-found` (#81803)
vercel/next.js: docs/01-app/03-api-reference/03-file-conventions/not-found.mdx

description: API reference for the not-found.js file.
---

Next.js provides two conventions to handle not found cases:

> - **`global-not-found.js`**: For your entire application.

Reviewer: Let's say it's for build the 404 page of the entire application. There're 2 phases in Next.js: "routing" and "rendering", if you're hitting a url that doesn't exist Next.js can render the global-not-found directly. And maybe only in global-not-found section so users won't need to re-think the existing `not-found.js` boundary

Author: Good nuance. 👍🏼

}
```

## `global-not-found.js` (experimental)

The `global-not-found.js` file lets you define a custom 404 page for your entire app, including its layout.

This is useful for:

- Apps with multiple root layouts.

> - Projects where composing a 404 page from `layout.js` and `not-found.js` is not possible.

Reviewer: these 2 are like the same cases, when you have multi root layouts, then it's not possible to choose one of them as the default root layout and find the closet `not-found.js` to build 404 page.

We can mention the "covered 2 cases" and "why":

- why: App Router has an implict convention where it compose root `layout.js` and `not-found.js` to as a 404 page, when routing results as 404, it will render that page. In few cases users will have hard time to generate the custom 404 page with the conventions due to the different project structure.

- 2 cases:
  - multi root layouts, e.g. multi group layouts
  - the layout and not found are not in the root project folder, e.g. using root params or flags at the top url segment.

To enable it, add the `globalNotFound` flag in `next.config.ts`:

```tsx filename="next.config.ts"
import type { NextConfig } from 'next'

const nextConfig: NextConfig = {
  experimental: {
    globalNotFound: true,
  },
}

export default nextConfig
```

Then, create a file in the root of the `app` directory: `app/global-not-found.js`:

```tsx filename="app/global-not-found.tsx" switcher
export default function GlobalNotFound() {
  return (
    <html>
      <body>
        <h1>404 - Page Not Found</h1>
        <p>This page does not exist.</p>
      </body>
    </html>
  )
}
```

Unlike `not-found.js`, this file must return a full HTML document, including `<html>` and `<body>` tags.

## Reference

### Props

`not-found.js` or `global-not-found.js` components do not accept any props.

> > **Good to know**: In addition to catching expected `notFound()` errors, the root `app/not-found.js` file handles any unmatched URLs for your whole application by default. This means users that visit a URL that is not handled by your app will be shown the UI exported by the `app/not-found.js` file. When the experimental `globalNotFound` flag is enabled, this behavior changes and the `app/global-not-found.js` file will handle unmatched URLs instead.

Author: ```suggestion
> **Good to know**: In addition to catching expected `notFound()` errors, the root `app/not-found.js` and `app/global-not-found.js` files handle any unmatched URLs for your whole application. This means users that visit a URL that is not handled by your app will be shown the exported UI.
```

If you need to use Client Component hooks like `usePathname` to display content based on the path, you must fetch data on the client-side instead.

### Metadata

> You can export a `metadata` object or a [`generateMetadata`](/docs/app/api-reference/functions/generate-metadata) function to customize the `<title>`, `<meta>`, and other head tags for your 404 page:

Author: Fake news.

> ```tsx filename="app/not-found.tsx" switcher

Reviewer: This `metadata` is currently only working smoothly with `global-not-found`, we haven't built the solid solution to locate `not-found.js`'s  metadata yet. Maybe only mention for global-not-found 🙏

export const metadata = {
  title: 'Not Found',
  description: 'The page you are looking for does not exist.',
}

> export default function NotFound() {

Reviewer: ```suggestion
export default function GlobalNotFound() {
```

  return (
    <div>
      <h1>Not Found</h1>
      <p>The page you are looking for does not exist.</p>
    </div>
  )
}
```

## Version History

| Version   | Changes                                             |
| --------- | --------------------------------------------------- |

> | `v15.x`   | `global-not-found.js` introduced (experimental).    |

Author: @huozhi what version was it introduced in?

Reviewer: Officially introduced in 15.4

Author: Ty!

> | `v15.x`   | `global-not-found.js` introduced (experimental).    |

Author: ```suggestion
| `v15.4.0`   | `global-not-found.js` introduced (experimental).    |
```
